ui/store/action-queue/index.js

```diff
@@ -0,0 +1,134 @@
+import pify from 'pify';
+import { isManifestV3 } from '../../../shared/modules/mv3.utils';
+
+// // A simplified pify maybe?
+// function pify(apiObject) {
+//   return Object.keys(apiObject).reduce((promisifiedAPI, key) => {
+//     if (apiObject[key].apply) { // depending on our browser support we might use a nicer check for functions here
+//       promisifiedAPI[key] = function (...args) {
+//         return new Promise((resolve, reject) => {
+//           return apiObject[key](
+//             ...args,
+//             (err, result) => {
+//               if (err) {
+//                 reject(err);
+//               } else {
+//                 resolve(result);
+//               }
+//             },
+//           );
+//         });
+//       };
+//     }
+//     return promisifiedAPI;
+//   }, {});
+// }
+
+let background = null;
+let promisifiedBackground = null;
+
+const actionRetryQueue = [];
+
+export function __TEST_CLEAR_QUEUE() {
+  actionRetryQueue.length = 0;
+}
+
+// add action to queue
+const addToActionQueue = (actionId, request, resolve, reject) => {
+  if (actionRetryQueue.some((act) => act.actionId === actionId)) {
+    return;
+  }
+  actionRetryQueue.push({
+    actionId,
+    request,
+    resolve,
+    reject,
+  });
+};
+
+// remove action from queue on successful completion
+// TODO: it should be possible to refactor the way the queue is drained to make this splicing unnecessary but I ran out of time.
+const removeFromActionQueue = (actionId) => {
+  const index = actionRetryQueue.find((act) => act.actionId === actionId);
+  actionRetryQueue.splice(index, 1);
+};
+
+// invokes promisifiedBackground method in MV2 content
+// In MV3 context the execution is:
+//   1. action is added to retry queue, along with resolve handler to be executed on completion of action
+//   2. is streams are connected promisifiedBackground method is called
+//   3. on completion of action either successfully or by throwing exception, action is removed from retry queue
+export const submitRequestToBackground = (
+  method,
+  args = [],
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
+) => {
+  if (isManifestV3()) {
+    return new Promise((resolve, reject) => {
+      addToActionQueue(actionId, { method, args }, resolve, reject);
+      drainActionRetryQueue();
```

Comment:
Can `drainActionRetryQueue` be part of `addToActionQueue` function itself ?

Reply:
yes. I wanted to do that and forgot.